File: 2022/day05/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

type Move struct {
	count int
	from  int
	to    int
}

func parseInput(filename string) ([][]byte, []Move) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// Find the blank line separating stacks from moves
	blankLine := 0
	for i, line := range lines {
		if line == "" {
			blankLine = i
			break
		}
	}

	stackLines := lines[:blankLine]
	moveLines := lines[blankLine+1:]

	// Find number of stacks from the last stack line (the numbers)
	numStacks := len(strings.Fields(stackLines[len(stackLines)-1]))

	// Parse stacks (excluding the number line)
	stacks := make([][]byte, numStacks)
	for i := range stacks {
		stacks[i] = []byte{}
	}

	for _, line := range stackLines[:len(stackLines)-1] {
		for i := 0; i < numStacks; i++ {
			pos := 1 + i*4 // Position of crate letter
			if pos < len(line) && line[pos] != ' ' {
				stacks[i] = append(stacks[i], line[pos])
			}
		}
	}

	// Reverse each stack so bottom is at index 0
	for i := range stacks {
		reverse(stacks[i])
	}

	// Parse moves
	moveRegex := regexp.MustCompile(`move (\d+) from (\d+) to (\d+)`)
	var moves []Move
	for _, line := range moveLines {
		if match := moveRegex.FindStringSubmatch(line); match != nil {
			count, _ := strconv.Atoi(match[1])
			from, _ := strconv.Atoi(match[2])
			to, _ := strconv.Atoi(match[3])
			moves = append(moves, Move{count, from - 1, to - 1}) // 0-indexed
		}
	}

	return stacks, moves
}

func reverse(s []byte) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

func copyStacks(stacks [][]byte) [][]byte {
	result := make([][]byte, len(stacks))
	for i, stack := range stacks {
		result[i] = make([]byte, len(stack))
		copy(result[i], stack)
	}
	return result
}

func part1(stacks [][]byte, moves []Move) string {
	stacks = copyStacks(stacks)
	for _, move := range moves {
		for i := 0; i < move.count; i++ {
			// Pop from source, push to destination
			crate := stacks[move.from][len(stacks[move.from])-1]
			stacks[move.from] = stacks[move.from][:len(stacks[move.from])-1]
			stacks[move.to] = append(stacks[move.to], crate)
		}
	}
	return topCrates(stacks)
}

func part2(stacks [][]byte, moves []Move) string {
	stacks = copyStacks(stacks)
	for _, move := range moves {
		// Move multiple crates at once (preserve order)
		fromLen := len(stacks[move.from])
		crates := make([]byte, move.count)
		copy(crates, stacks[move.from][fromLen-move.count:])
		stacks[move.from] = stacks[move.from][:fromLen-move.count]
		stacks[move.to] = append(stacks[move.to], crates...)
	}
	return topCrates(stacks)
}

func topCrates(stacks [][]byte) string {
	var result []byte
	for _, stack := range stacks {
		if len(stack) > 0 {
			result = append(result, stack[len(stack)-1])
		}
	}
	return string(result)
}

func main() {
	_, currentFile, _, _ := runtime.Caller(0)
	dir := filepath.Dir(currentFile)
	inputFile := filepath.Join(dir, "..", "input.txt")

	stacks, moves := parseInput(inputFile)

	fmt.Println("Part 1:", part1(stacks, moves))
	fmt.Println("Part 2:", part2(stacks, moves))
}

File: 2022/day05/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <regex>

struct Move {
    int count;
    int from;
    int to;
};

std::pair<std::vector<std::vector<char>>, std::vector<Move>> parseInput(const std::string& filename) {
    std::ifstream file(filename);
    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());

    // Split by double newline
    size_t splitPos = content.find("\n\n");
    std::string stackPart = content.substr(0, splitPos);
    std::string movePart = content.substr(splitPos + 2);

    // Parse stack lines
    std::vector<std::string> stackLines;
    std::istringstream stackStream(stackPart);
    std::string line;
    while (std::getline(stackStream, line)) {
        stackLines.push_back(line);
    }

    // Find number of stacks from the last line
    std::string numberLine = stackLines.back();
    int numStacks = 0;
    std::istringstream numStream(numberLine);
    int num;
    while (numStream >> num) {
        numStacks++;
    }

    // Parse stacks (excluding the number line)
    std::vector<std::vector<char>> stacks(numStacks);
    for (size_t i = 0; i < stackLines.size() - 1; i++) {
        const std::string& stackLine = stackLines[i];
        for (int j = 0; j < numStacks; j++) {
            int pos = 1 + j * 4;  // Position of crate letter
            if (pos < static_cast<int>(stackLine.size()) && stackLine[pos] != ' ') {
                stacks[j].push_back(stackLine[pos]);
            }
        }
    }

    // Reverse so bottom is at index 0
    for (auto& stack : stacks) {
        std::reverse(stack.begin(), stack.end());
    }

    // Parse moves
    std::vector<Move> moves;
    std::regex moveRegex(R"(move (\d+) from (\d+) to (\d+))");
    std::istringstream moveStream(movePart);
    while (std::getline(moveStream, line)) {
        std::smatch match;
        if (std::regex_match(line, match, moveRegex)) {
            Move m;
            m.count = std::stoi(match[1]);
            m.from = std::stoi(match[2]) - 1;  // 0-indexed
            m.to = std::stoi(match[3]) - 1;
            moves.push_back(m);
        }
    }

    return {stacks, moves};
}

std::string part1(std::vector<std::vector<char>> stacks, const std::vector<Move>& moves) {
    for (const auto& move : moves) {
        for (int i = 0; i < move.count; i++) {
            char crate = stacks[move.from].back();
            stacks[move.from].pop_back();
            stacks[move.to].push_back(crate);
        }
    }

    std::string result;
    for (const auto& stack : stacks) {
        if (!stack.empty()) {
            result += stack.back();
        }
    }
    return result;
}

std::string part2(std::vector<std::vector<char>> stacks, const std::vector<Move>& moves) {
    for (const auto& move : moves) {
        // Move multiple crates at once (preserve order)
        std::vector<char> crates(stacks[move.from].end() - move.count, stacks[move.from].end());
        stacks[move.from].resize(stacks[move.from].size() - move.count);
        stacks[move.to].insert(stacks[move.to].end(), crates.begin(), crates.end());
    }

    std::string result;
    for (const auto& stack : stacks) {
        if (!stack.empty()) {
            result += stack.back();
        }
    }
    return result;
}

int main() {
    auto [stacks, moves] = parseInput("../input.txt");

    std::cout << "Part 1: " << part1(stacks, moves) << std::endl;
    std::cout << "Part 2: " << part2(stacks, moves) << std::endl;

    return 0;
}
